### PR TITLE
Create explicit Token constructors, Token and Marker as struct

### PR DESCRIPTION
**Pull request by lahma**
* factory constructors initialize use case specific fields
* trim Token fields to minimize count and size
* change Token to be a read-only struct

Jint tests are unaffected by this.

### CODE REVIEW

**Comment by jogibear9988** [comment]
Is a struct really preferable?

When I read here:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/choosing-between-class-and-struct

- It has an instance size under 16 bytes.

If I look at your code:

        public readonly struct Token
        {
            private Token(
                TokenType type,
                object? value,
                int start,
                int end,
                int lineNumber, 
                int lineStart,
                bool octal = false,
                char notEscapeSequenceHead = (char) 0,
                bool head = false,
                bool tail = false, 
                object? customValue = null)

these are 11 Properties, 4x an int32 wich are already 16 bytes, two pointers = 8 or 16bytes, 3 bools, 1 enum, 1char = 5bytes, 
so we have already 29bytes (or 37 if a pointer is 8 byte)

**Comment by lahma** [comment]
Here are the first results for struct-based solution.

``` ini
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22621
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=6.0.400
  [Host]     : .NET 6.0.8 (6.0.822.36306), X64 RyuJIT
  DefaultJob : .NET 6.0.8 (6.0.822.36306), X64 RyuJIT
```

## Esprima.Benchmark.FileParsingBenchmark

| **Diff**|Method|FileName|Mean|Error|Gen 0|Gen 1|Gen 2|Allocated|
|------- |-------|-------|-------:|-------|-------:|-------:|-------:|-------:|
| Old |ParseProgram|angular-1.2.5|30.478 ms|0.3480 ms|1687.5000|968.7500|343.7500|24 MB|
| **New** |	|	| **26.157 ms (-14%)** | **0.2075 ms** | **1000.0000 (-41%)** | **562.5000 (-42%)** | **187.5000 (-45%)** | **14 MB (-42%)** |
| Old |ParseProgram|backbone-1.1.0|3.167 ms|0.0034 ms|257.8125|136.7188|54.6875|4 MB|
| **New** |	|	| **3.143 ms (-1%)** | **0.0310 ms** | **160.1563 (-38%)** | **93.7500 (-31%)** | **31.2500 (-43%)** | **2 MB (-50%)** |
| Old |ParseProgram|jquery-1.9.1|23.224 ms|0.2257 ms|1406.2500|843.7500|312.5000|20 MB|
| **New** |	|	| **20.214 ms (-13%)** | **0.2625 ms** | **843.7500 (-40%)** | **500.0000 (-41%)** | **187.5000 (-40%)** | **12 MB (-40%)** |
| Old |ParseProgram|jquery.mobile-1.4.2|46.191 ms|0.8820 ms|2272.7273|1363.6364|454.5455|33 MB|
| **New** |	|	| **33.333 ms (-28%)** | **0.4756 ms** | **1266.6667 (-44%)** | **733.3333 (-46%)** | **333.3333 (-27%)** | **19 MB (-42%)** |
| Old |ParseProgram|mootools-1.4.5|14.784 ms|0.1339 ms|1109.3750|578.1250|234.3750|17 MB|
| **New** |	|	| **15.343 ms (+4%)** | **0.0195 ms** | **703.1250 (-37%)** | **421.8750 (-27%)** | **203.1250 (-13%)** | **10 MB (-41%)** |
| Old |ParseProgram|underscore-1.5.2|1.419 ms|0.0037 ms|199.2188|99.6094|-|3 MB|
| **New** |	|	| **1.541 ms (+9%)** | **0.0055 ms** | **117.1875 (-41%)** | **56.6406 (-43%)** | **-** | **2 MB (-33%)** |
| Old |ParseProgram|yui-3.12.0|17.831 ms|0.2276 ms|1093.7500|625.0000|218.7500|16 MB|
| **New** |	|	| **13.111 ms (-26%)** | **0.0965 ms** | **671.8750 (-39%)** | **437.5000 (-30%)** | **203.1250 (-7%)** | **9 MB (-44%)** |

@jogibear9988 I believe the documented rules are for generic types passed around, like `DateTime` etc. In this case Token is very short lived and a lot of them are being created and then just tossed away. The numbers pretty much speak for themselves I guess.

**Comment by lahma** [comment]
@adams85 please take a look, I know that the `TokenType` is now mixing again JSX concepts, but this solution keeps the struct small and manageable without tricks.

File: src/Esprima/Ast/Literal.cs [hunk 1]
```diff
@@ -48,7 +56,7 @@ public Literal(string pattern, string flags, object? value, string raw) : this(T
     public RegexValue? Regex { [MethodImpl(MethodImplOptions.AggressiveInlining)] get; }
 
     public string? StringValue { [MethodImpl(MethodImplOptions.AggressiveInlining)] get => TokenType == TokenType.StringLiteral ? Value as string : null; }
-    public bool BooleanValue { [MethodImpl(MethodImplOptions.AggressiveInlining)] get => TokenType == TokenType.BooleanLiteral && NumericValue != 0; }
+    public bool BooleanValue { [MethodImpl(MethodImplOptions.AggressiveInlining)] get => TokenType == TokenType.BooleanLiteral && ReferenceEquals(Value, _boxedTrue); }
```
Comment (jogibear9988) [inline src/Esprima/Ast/Literal.cs:59]: False is no boolean Value?

Comment (lahma) [reply src/Esprima/Ast/Literal.cs:59]: On this case it should be enough to check only for true being set, like old logic checked for non-zero

**Comment by lahma** [comment]
I also converted Marker to struct and some `in` usage to avoid copies, which gave further improvements. The benchmark has collect tokens = true, which creates unnecessary red herrings about memory usage (now that `ParsedToken` is separate, it was wasy to spot), but should be switched to off as a separate task.

## Esprima.Benchmark.FileParsingBenchmark

| **Diff**|Method|FileName|Mean|Error|Gen 0|Gen 1|Gen 2|Allocated|
|------- |-------|-------|-------:|-------|-------:|-------:|-------:|-------:|
| Old |ParseProgram|angular-1.2.5|30.478 ms|0.3480 ms|1687.5000|968.7500|343.7500|24 MB|
| **New** |	|	| **24.277 ms (-20%)** | **0.1647 ms** | **937.5000 (-44%)** | **562.5000 (-42%)** | **187.5000 (-45%)** | **13 MB (-46%)** |
| Old |ParseProgram|backbone-1.1.0|3.167 ms|0.0034 ms|257.8125|136.7188|54.6875|4 MB|
| **New** |	|	| **2.939 ms (-7%)** | **0.0333 ms** | **144.5313 (-44%)** | **93.7500 (-31%)** | **31.2500 (-43%)** | **2 MB (-50%)** |
| Old |ParseProgram|jquery-1.9.1|23.224 ms|0.2257 ms|1406.2500|843.7500|312.5000|20 MB|
| **New** |	|	| **19.963 ms (-14%)** | **0.3915 ms** | **750.0000 (-47%)** | **437.5000 (-48%)** | **156.2500 (-50%)** | **11 MB (-45%)** |
| Old |ParseProgram|jquery.mobile-1.4.2|46.191 ms|0.8820 ms|2272.7273|1363.6364|454.5455|33 MB|
| **New** |	|	| **25.223 ms (-45%)** | **0.3831 ms** | **1312.5000 (-42%)** | **843.7500 (-38%)** | **406.2500 (-11%)** | **17 MB (-48%)** |
| Old |ParseProgram|mootools-1.4.5|14.784 ms|0.1339 ms|1109.3750|578.1250|234.3750|17 MB|
| **New** |	|	| **14.424 ms (-2%)** | **0.1609 ms** | **640.6250 (-42%)** | **421.8750 (-27%)** | **203.1250 (-13%)** | **9 MB (-47%)** |
| Old |ParseProgram|underscore-1.5.2|1.419 ms|0.0037 ms|199.2188|99.6094|-|3 MB|
| **New** |	|	| **1.454 ms (+2%)** | **0.0059 ms** | **105.4688 (-47%)** | **35.1563 (-65%)** | **-** | **2 MB (-33%)** |
| Old |ParseProgram|yui-3.12.0|17.831 ms|0.2276 ms|1093.7500|625.0000|218.7500|16 MB|
| **New** |	|	| **13.457 ms (-25%)** | **0.2148 ms** | **625.0000 (-43%)** | **437.5000 (-30%)** | **203.1250 (-7%)** | **8 MB (-50%)** |

**Review by adams85 — COMMENTED** [review]
Awesome improvement! The benchmark result speaks for itself: even execution time is reduced despite the increased amount of copying. (I mean there are a bunch of functions in scanner and parser returning `Token` which implies copying the whole from now on vs. copying a pointer-sized reference before. Thus, there may be some further room for improvement: instead of returning `Token`, functions could take a `ref Token`/`out Token` parameter. It would be interesting to see what impact this would have on execution time.)

Another thing worth mentioning is that the `Token`/`ParsedToken` separation should be done in the case of `Comment` as well. I've already thought of converting `Comment` to a value type but I'd rather have waited for this PR to make it consistent with tokens.

File: src/Esprima/Scanner.cs [hunk 2]
```diff
@@ -707,20 +707,12 @@ private Token ScanIdentifier(bool allowEscapes)
...
-            Start = start,
-            End = Index
-        };
+        return Token.Create(type, id, start, end: Index, LineNumber, LineStart);
     }
 
     // https://tc39.github.io/ecma262/#sec-punctuators
 
-    public Token ScanPunctuator()
+    private Token ScanPunctuator()
```
[the remark this answers is not in the document]
Comment (lahma) [reply src/Esprima/Scanner.cs:715]: Good catch, changed them to private/internal too.

File: src/Esprima/Token.cs [hunk 3]
```diff
@@ -18,40 +21,117 @@ public enum TokenType
     Template,
     BigIntLiteral,
 
-    Extension = int.MaxValue
-};
+    JsxIdentifier,
+    JsxText,
+    Extension
```
[the remark this answers is not in the document]
Comment (lahma) [reply src/Esprima/Token.cs:26]: Great, I'll leave as-is and see what future holds.

File: src/Esprima/Token.cs [hunk 4]
```diff
@@ -18,40 +21,117 @@ public enum TokenType
...
-    public object? Value;
-    public RegexValue? RegexValue;
-    public BigInteger? BigIntValue;
+    internal static Token CreatePunctuator(string str, int start, int end, int lineNumber, int lineStart)
+    {
+        return new Token(TokenType.Punctuator, str, start, end, lineNumber, lineStart);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    internal static Token CreteTemplate(
```
[the remark this answers is not in the document]
Comment (lahma) [reply src/Esprima/Token.cs:99]: Fixed

**Comment by lahma** [comment]
Thanks for the review, at least I got something completed!

**Pull request merged**